File: core/src/main/java/com/arsvechkarev/core/WordsDatabase.java
package com.arsvechkarev.core;

import android.content.Context;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import com.arsvechkarev.core.domain.dao.WordDao;
import com.arsvechkarev.core.domain.model.Word;

@Database(
    entities = {Word.class},
    version = 1
)
public abstract class WordsDatabase extends RoomDatabase{

  public abstract WordDao wordDao();

  private static WordsDatabase instance;

  public static void instantiate(Context context) {
    instance = Room.databaseBuilder(
        context,
        WordsDatabase.class,
        "words.db"
    ).build();
  }


  public static WordsDatabase getInstance() {
    return instance;
  }
}

File: storage/src/main/java/com/arsvechkarev/storage/database/WordsDatabase.kt
package com.arsvechkarev.storage.database

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.arsvechkarev.core.domain.dao.LabelsDao
import com.arsvechkarev.core.domain.dao.WordsDao
import com.arsvechkarev.core.domain.dao.WordsLabelsDao
import com.arsvechkarev.core.domain.model.LabelEntity
import com.arsvechkarev.core.domain.model.WordEntity

@Database(entities = [
  WordEntity::class,
  LabelEntity::class,
  WordsLabelsDao::class
], version = 1)
abstract class WordsDatabase : RoomDatabase() {
  
  abstract fun wordDao(): WordsDao
  abstract fun labelsDao(): LabelsDao
  abstract fun wordsAndLabelsDao(): WordsLabelsDao
  
  companion object {
    
    lateinit var database: WordsDatabase
      private set
    
    lateinit var wordsDao: WordsDao
      private set
    
    fun instantiate(context: Context) {
      database = Room.databaseBuilder(
        context,
        WordsDatabase::class.java,
        "words.db"
      ).build()
      wordsDao = database.wordDao()
    }
  }
}
